Enable page size in campaigns endpoint

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Part of pcTzPl-2nS-p2

This PR allows setting the per_page parameter in the gla/ads/campaigns endpoint. Previously, we returned all the results, but now it is possible to limit the number of results.


### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Make a request to `GET gla/ads/campaigns?per_page=X` and check that the number of results matches the per_page parameter.
2. Make a request to `GET gla/ads/campaigns` and ensure that you get all the results.

### Additional details:

I had to replace [iterateAllElements()](https://googleapis.github.io/gax-php/master/Google/ApiCore/PagedListResponse.html#method_iterateAllElements) with `getPage()->getIterator()` to avoid iterating through all the pages and retrieving all the results at once.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Update - Enable Page Size Parameter in Campaigns Endpoint

jorgemd24 committed:
Allow per_page parameter in campaign endpoint

jorgemd24 committed:
Add per_page CampaignController parameter

jorgemd24 committed:
Adjust tests

jorgemd24 committed:
PHPcs tests

jorgemd24 committed:
phpcs tests

puntope commented on src/API/Google/AdsCampaign.php:
```diff
@@ -99,15 +99,16 @@ public function __construct( GoogleAdsClient $client, AdsCampaignBudget $budget,
-	 * @param bool $fetch_criterion Combine the campaign data with criterion data (default true).
+	 * @param bool  $exclude_removed Exclude removed campaigns (default true).
+	 * @param bool  $fetch_criterion Combine the campaign data with criterion data (default true).
+	 * @param array $args Additional arguments.
```
WDYT of being more concrete here?

Maybe 

 @param array $args Arguments for the Ads Campaign Query.

puntope commented on tests/Unit/API/Google/AdsCampaignTest.php:
```diff
@@ -136,6 +136,66 @@ public function test_get_campaigns() {
+				'campaign_id'         => self::TEST_CAMPAIGN_ID,
+				'geo_target_constant' => 'geoTargetConstants/2158',
+			],
+			[
+				'campaign_id'         => 5678901234,
+				'geo_target_constant' => 'geoTargetConstants/2344',
+			],
+			[
+				'campaign_id'         => 5678901234,
+				'geo_target_constant' => 'geoTargetConstants/2826',
+			],
```
Its just a test, but the ids are duplicated.

puntope commented on tests/Unit/API/Site/Controllers/Ads/CampaignControllerTest.php:
```diff
@@ -152,6 +152,67 @@ public function test_get_campaigns_converted_names() {
 		$this->assertEquals( 200, $response->get_status() );
 	}
 
+	public function test_get_campaigns_with_args() {
```
In the test added here where are we testing another page_size? For example page_size=1 to see if we don't get all the campaigns and just 1

puntope reviewed (commented):
Thanks for the changes @jorgemd24 

I left some minor comments but the per_page works as expected.

One thing I miss is the page parameter to get the results for a specific page. Not sure if its for an other PR.

jorgemd24 committed:
Adjust phpdoc

jorgemd24 committed:
Test different page size

jorgemd24 committed:
Add pagination header for the campaign endpoint

jorgemd24 committed:
Adjust tests to check pagination headers

jorgemd24 commented:
Thanks @puntope for reviewing this PR! I addressed your suggestions.

> One thing I miss is the page parameter to get the results for a specific page. Not sure if its for an other PR.

Initially, I planned not to set the next page token in the endpoint response for the reasons mentioned in this comment: https://github.com/woocommerce/google-listings-and-ads/issues/1887#issuecomment-2231712265. However, after @mikkamp's comment, I realized it is possible to get the total number of results in the first query. Therefore, I implemented the following headers:

- `X-WP-Total` -> Total number across all pages.
- `X-WP-TotalPages` -> Total pages.
- `X-GLA-NextPageToken` -> The token to get the next page. (I used the GLA prefix to avoid future conflicts with this header name)

With this, we can fetch the next page and store the page token in the client to navigate backwards. However, it seems that it is not possible to get the forward page tokens until you access the page, (for example jumping from page 1 to page 5) so we need to see if it is feasible to implement this in the front end and the back end. 

I found this example that can be useful: https://developers.google.com/google-ads/api/samples/navigate-search-result-pages-caching-tokens but for now, I will leave that implementation/investigation for a different PR as the mobile team is interested in limiting the result to 1 to check if the merchant is active and I think this PR will help us to move forward to implement the pagination in the campaign table.

puntope commented on src/API/Google/AdsCampaign.php:
```diff
@@ -99,15 +99,17 @@ public function __construct( GoogleAdsClient $client, AdsCampaignBudget $budget,
 	 * @throws ExceptionWithResponseData When an ApiException is caught.
 	 */
-	public function get_campaigns( bool $exclude_removed = true, bool $fetch_criterion = true ): array {
+	public function get_campaigns( bool $exclude_removed = true, bool $fetch_criterion = true, $args = [], $return_pagination_params = false ): array {
```
What's the value of NOT returning pagination params? Even if they are not needed I assume they are not harmful. So maybe we can simplify this method by returning them always.

jorgemd24 replied:
@puntope, the reason for not always returning the pagination params is that we would need to refactor other parts of the code and tests that depend on this method:

- https://github.com/woocommerce/google-listings-and-ads/blob/6ff59162df0fc9c79e1f0c1af8267a764fe143a9/src/API/Google/AdsCampaign.php#L405
- https://github.com/woocommerce/google-listings-and-ads/blob/6ff59162df0fc9c79e1f0c1af8267a764fe143a9/src/MultichannelMarketing/GLAChannel.php#L210
- https://github.com/woocommerce/google-listings-and-ads/blob/6ff59162df0fc9c79e1f0c1af8267a764fe143a9/tests/Unit/API/Google/AdsCampaignTest.php#L95
- ..etc

Do you think it’s worth refactoring those parts even if they don't require pagination?

puntope reviewed (approved):
Thx. @jorgemd24 I think we can go forward with this and evaluate the refactoring you mentioned
